Add initial draft of Podman project Governance (#25398)
containers/podman: REVIEWING.md

All code merged must pass CI.

## What should you review?

We encourage review of all PRs, even those not in the area of expertise of the reviewer.
Even if you are not fully familiar with the code in question, you can still notice basic mistakes.
Feel free to ask questions about how areas of the code work to help familiarize yourself during reviews.
If you finish a review and do not feel like you adequately understood the code to approve it for merge, tag an expert in that area of the code to perform a further review.

Timely PR reviews are important - contributors can become discouraged if a PR is neglected by maintainers.
All Maintainers and Reviewers should try to review new pull requests in their repositories once a day to ensure this.
When you review a PR with failing tests, please check to see if those tests failed due to known flakes.
If so, please restart the failed tests.
Many repositories in the Podman project can only have their tests restarted by project members, not the submitter, so regular attention to test failures by reviewers is important.

## Things to Check

### Breaking Changes

The Podman project aims to present a stable API for its users.

> Breaking changes to the project's Command Line Interfaces or public APIs (the Podman REST API and its associated bindings) must only be made in approved breaking change releases - Podman 6.0, 7.0, etc.

Reviewer: I think it would be great to add a list of directories where the stable parts of the project are located. So that new maitainers would know that they are changing a stable part and possibly avoid it, and so that new reviewers could check these places more carefully. I don't know if the PR label is sufficient when you're interfering with multiple parts of a project.

Breaking changes can include renaming an option without retaining the original name as an alias, removing an option entirely, or changing how an option works in a way that does not ensure backwards compatibility.

Periods when it is acceptable to merge breaking changes will be widely announced.
If it is not one of those periods, reviewers should be on the lookout for breaking changes.
PRs with breaking changes should not be merged.
Please guide the contributor in how to make the change in a non-breaking fashion.
If this is not possible, deferring the PR to the next breaking change window or closing it entirely is appropriate.

### Commit Messages and Hygiene

Good commit messages are essential for understanding why a change was made in the future.
Reviewers should check each commit of a PR to ensure that it has an appropriate commit message which fully and accurately explains the change and why it is being made.

> This should remain true after changes to the PR due to code review, so please re-review commit messages before merging code, to ensure they are still accurate.

Reviewer 2: Commit messages must have `Fixes: #xxxx` with the right issue number if they claim to fix said issue.

If there is a disagreement, please do not make the contributor make repeated changes until an agreement on how to proceed is reached.
Also, please attempt to reach an agreement quickly, so the PR can be merged in a timely fashion.

### Language Version Updates

Most repositories in the Podman project are written in Go, and as such target a specific version of Go in their `go.mod`.
For example:
```
$ cat go.mod | grep 'go 1.'
go 1.22.8
```

For all branches except the main branch, this should always remain static.
PRs into a non-main branch which change the supported Go version should be modified to not require such a change, or rejected if that is not possible.
Changing supported Go version in the main branch is allowed, but not encouraged.
Doing so will require significant discussion among maintainers, as it can affect what distributions the main branch can be built.

### Tests

Please Check tests added by the PR.
If a PR has no new tests, determine if this is actually appropriate - has new functionality been added which should have been tested?
If the PR does have tests, check to see if they are reasonably comprehensive.

> The Podman Project does not have code coverage standards at present, but we aim to ensure that all new functionality is tested.

Reviewer 2: Also make sure the newly added test for a bug fix/feature fails against a binary from current main. If that is not the case the test is not good as it does not check the newly added change. As such it should be modified until that this is the case.

### Documentation

All changes to public-facing APIs (e.g. the Podman REST API) and CLI should be appropriately documented.
API changes require Swagger documentation.
CLI changes should be documented in the Manpages.

Please validate that PRs making such changes include appropriate documentation.
Many repositories in the Podman project will enforce this via CI check, but you should still review the contents of the documentation to make sure they are appropriate and complete.

## Things to Avoid

### Bikeshedding and excessively critical reviews

Please avoid bikeshedding during reviews.

Trivial changes that do not affect the ultimate functionality of the PR - for example, small grammar or phrasing changes in documentation, unnecessary renaming of variables - should not block merge of a PR.
It is acceptable to make such comments, but they should be marked as nice to have changes.

### Asking for unrelated changes

> Please do not request that a contributor make significant changes to code their PR did not touch.

Reviewer 2: I understand the intention but I am not sure that is the right outcome, some changes might absolutely require some form of larger rework maybe in places they did not touch.
Otherwise you accept hack after hack and the code ends up in a total mess. So if the code is changing something of significant impact and does not seem to be the best implementation maintenance wise then I think it must be valid to block on this.
The end goal of code review should also be to keep the code base maintainable. I do acknowledge what is "maintainable" is totally up to the individual maintainer/reviewer and everybody will have different opinions.

Author: Maybe I should put something in here about "best judgement" - relatively minor refactoring is probably fine, but I don't want anyone requesting wide, sweeping changes that will double the size of a PR and introduce a major burden on the contributor.

Author: Added some sentences to that effect

Reviewer 2: Thanks, new text sounds good to me.

# Reviewing Pull Requests

This document contains general principles for how to perform code reviews in the Podman repository.
It does not aim to be a complete guide to how to perform code review, but rather to provide general guidance on how code reviews should be performed.

> This document is aimed at Reviewers, Maintainers, and Core Maintainers (see [GOVERNANCE.md](./GOVERNANCE.md) for definitions of these roles), but these guidelines shpould be followed by all who wish to review code in the Podman project's Github repositories, even those who are not currently a maintainer.

Reviewer 3: ```suggestion
This document is aimed at Reviewers, Maintainers, and Core Maintainers (see [GOVERNANCE.md](./GOVERNANCE.md) for definitions of these roles), but these guidelines should be followed by all who wish to review code in the Podman project's Github repositories, even those who are not currently a maintainer.
```

All code merged must pass CI.

## What should you review?

We encourage review of all PRs, even those not in the area of expertise of the reviewer.
Even if you are not fully familiar with the code in question, you can still notice basic mistakes.
Feel free to ask questions about how areas of the code work to help familiarize yourself during reviews.
If you finish a review and do not feel like you adequately understood the code to approve it for merge, tag an expert in that area of the code to perform a further review.

Timely PR reviews are important - contributors can become discouraged if a PR is neglected by maintainers.
All Maintainers and Reviewers should try to review new pull requests in their repositories once a day to ensure this.
When you review a PR with failing tests, please check to see if those tests failed due to known flakes.
If so, please restart the failed tests.
Many repositories in the Podman project can only have their tests restarted by project members, not the submitter, so regular attention to test failures by reviewers is important.

## Things to Check

### Breaking Changes

The Podman project aims to present a stable API for its users.

> Breaking changes to the project's Command Line Interfaces or public APIs (the Podman REST API) must only be made in approved breaking change releases - Podman 6.0, 7.0, etc.

Reviewer 4: I think pkg/bindings should be stable as well?

Disagreements between maintainers can generally be worked out during PR review through comments.
If this is not possible, other maintainers can be called on to give their opinions and determine a way forward.
We encourage such disagreements and discussions, so long as they remain respectful.

Being respectful includes respect for the contributor's time.
If there is a disagreement, please do not make the contributor make repeated changes until an agreement on how to proceed is reached.
Also, please attempt to reach an agreement quickly, so the PR can be merged in a timely fashion.

### Language Version Updates

Most repositories in the Podman project are written in Go, and as such target a specific version of Go in their `go.mod`.
For example:
```
$ cat go.mod | grep 'go 1.'
go 1.22.8
```

For all branches except the main branch, this should always remain static.
PRs into a non-main branch which change the supported Go version should be modified to not require such a change, or rejected if that is not possible.
Changing supported Go version in the main branch is allowed, but not encouraged.

> Doing so will require significant discussion among maintainers, as it can effect on what distributions the main branch can be built.

Reviewer 3: ```suggestion
Doing so will require significant discussion among maintainers, as it can affect on what distributions the main branch can be built.
```

### Tests

Please Check tests added by the PR.
If a PR has no new tests, determine if this is actually appropriate - has new functionality been added which should have been tested?
If the PR does have tests, check to see if they are reasonably comprehensive.
The Podman Project does not have code coverage standards at present, but we aim to ensure that all new functionality is tested.

### Documentation

All changes to public-facing APIs (e.g. the Podman REST API) and CLI should be appropriately documented.
API changes require Swagger documentation.
CLI changes should be documented in the Manpages.

Please validate that PRs making such changes include appropriate documentation.
Many repositories in the Podman project will enforce this via CI check, but you should still review the contents of the documentation to make sure they are appropriate and complete.

## Things to Avoid

### Bikeshedding and excessively critical reviews

Please avoid bikeshedding during reviews.

> Trivial changes that do not effect the ultimate functionality of the PR - for example, small grammer or phrasing changes in documentation, unnecessary renaming of variables - should not block merge of a PR.

Reviewer 3: ```suggestion
Trivial changes that do not affect the ultimate functionality of the PR - for example, small grammar or phrasing changes in documentation, unnecessary renaming of variables - should not block merge of a PR.
```

It is acceptable to make such comments, but they should be marked as nice to have changes.

Also, it should not have to be said, but please be polite during code reviews.

> The [CNCF Code of Conduct](https://github.com/cncf/foundation/blob/main/code-of-conduct.md) applies to all reviewers, of course, but in order to encourage repeat contributions, reviewers should be polite and pleasant when performing reviews.

Reviewer 3: Politeness can be hard to judge given the same language is spoken differently around the world. Can we add `assume good intent` here? Also, if there's a recommended list of words to avoid that can be linked here, that'd be nice I guess.

Author: Sure

All code merged must pass CI.

## What should you review?

We encourage review of all PRs, even those not in the area of expertise of the reviewer.
Even if you are not fully familiar with the code in question, you can still notice basic mistakes.
Feel free to ask questions about how areas of the code work to help familiarize yourself during reviews.
If you finish a review and do not feel like you adequately understood the code to approve it for merge, tag an expert in that area of the code to perform a further review.

Timely PR reviews are important - contributors can become discouraged if a PR is neglected by maintainers.
All Maintainers and Reviewers should try to review new pull requests in their repositories once a day to ensure this.
When you review a PR with failing tests, please check to see if those tests failed due to known flakes.
If so, please restart the failed tests.
Many repositories in the Podman project can only have their tests restarted by project members, not the submitter, so regular attention to test failures by reviewers is important.

## Things to Check

### Breaking Changes

The Podman project aims to present a stable API for its users.

> Breaking changes to the project's Command Line Interfaces or public APIs (the Podman REST API and its associated bindings) must only be made in approved breaking change releases - Podman 6.0, 7.0, etc.

Reviewer 5: I think there are precedents for exceptions that should possibly be noted OR a more general statement saying exceptions may be made by whatever/whoever.  Things I am thinking of:

* correction of a bug where it was implemented incorrectly . Example would be docker compat
* security related
* altering of defaults specifically for the benefit of users

Reviewer 6: That's an excellent point, especially the security angle....

Author: If we had to do that, it would have to be Podman 6.0. We don't do breaking changes in 5.x, even for security.

Individual repositories should identify what parts of the repository are considered to be their stable API.
Breaking changes can include renaming an option without retaining the original name as an alias, removing an option entirely, or changing how an option works in a way that does not ensure backwards compatibility.

Periods when it is acceptable to merge breaking changes will be widely announced.
If it is not one of those periods, reviewers should be on the lookout for breaking changes.
PRs with breaking changes should not be merged.
Please guide the contributor in how to make the change in a non-breaking fashion.
If this is not possible, deferring the PR to the next breaking change window or closing it entirely is appropriate.

### Commit Messages and Hygiene

Good commit messages are essential for understanding why a change was made in the future.
Reviewers should check each commit of a PR to ensure that it has an appropriate commit message which fully and accurately explains the change and why it is being made.
This should remain true after changes to the PR due to code review, so please re-review commit messages before merging code, to ensure they are still accurate.

> Pull requests fixing a specific issue must include a `Fixes: #xxxxx` line in the commit message.

Reviewer 5: in the case of Jira, I think @Luap99 recommends the plain link instead of a GH reference.  @Luap99 what say you?

Reviewer 2: Sure but IMO this document should not need duplicate what is written in CONTRIBUTING.md, i.e.
https://github.com/containers/podman/blob/main/CONTRIBUTING.md#describe-your-changes-in-commit-messages

Overall I think most part should be ensure CONTRIBUTING.md is followed, this here doc here should then just help is small review guidance so people don't forget to double check these things but there should be no need to go into the full details here again

Each commit in a PR should be self-contained and have a clear and distinct purpose.
If this is not true, please encourage the contributor to squash their commits using `git rebase -i` until it is true.

### Disagreements between reviewers

We do not expect all reviewers to be of the same opinion during code review.
If you see another reviewer requesting changes to a PR you do not agree with, it is perfectly acceptable to comment to that effect.
Disagreements between maintainers can generally be worked out during PR review through comments.
If this is not possible, other maintainers can be called on to give their opinions and determine a way forward.

> We encourage such disagreements and discussions, so long as they remain respectful.

Reviewer 5: ```suggestion
We encourage such disagreements and discussions, so long as they remain respectful and are working towards the goal of dispute resolution.
```

Reviewer 7: ```suggestion
We encourage such disagreements and discussions, so long as they remain respectful.

Final decisions on the project reside with the Core Maintainers. If the Core Maintainers cannot decide an issue, the issue will be resolved by voting. The voting process is based on a [super-majority](https://en.wikipedia.org/wiki/Supermajority#Two-thirds_vote) vote where each Core Maintainers is given one voice.
```

# Reviewing Pull Requests

This document contains general principles for how to perform code reviews in the Podman repository.
It does not aim to be a complete guide to how to perform code review, but rather to provide general guidance on how code reviews should be performed.

This document is aimed at Reviewers, Maintainers, and Core Maintainers (see [GOVERNANCE.md](./GOVERNANCE.md) for definitions of these roles), but these guidelines should be followed by all who wish to review code in the Podman project's Github repositories, even those who are not currently a maintainer.

## How are reviews performed

The Podman project aims to ensure that all PRs are reviewed by at least 2 people prior to merge, at least one of which must be a repository Maintainer.
There are some exceptions to this: Updates to libraries (including Go vendor updates) that pass CI cleanly and require no code changes may be merged by a maintainer without further review.

> All code merged must pass CI.

Author: That's enforced by CI, no need for reviewers to pay particular attention

If you see another reviewer requesting changes to a PR you do not agree with, it is perfectly acceptable to comment to that effect.
Disagreements between maintainers can generally be worked out during PR review through comments.
If this is not possible, other maintainers can be called on to give their opinions and determine a way forward.
We encourage such disagreements and discussions, so long as they remain respectful and are done with the goal of resolving the dispute.

If a decision cannot be reached, the issue may be put to a vote, in which all Maintainers of the repository in question and all Core Maintainers can vote.

Being respectful includes respect for the contributor's time.
If there is a disagreement, please do not make the contributor make repeated changes until an agreement on how to proceed is reached.
Also, please attempt to reach an agreement quickly, so the PR can be merged in a timely fashion.

### Language Version Updates

Most repositories in the Podman project are written in Go, and as such target a specific version of Go in their `go.mod`.
For example:
```
$ cat go.mod | grep 'go 1.'
go 1.22.8
```

> For all branches except the main branch, this should always remain static.

Reviewer 6: "should" is probably a good word here.  We may end up bumping the go version in the go.mod files of some of the RHEL branches to resolve the recent crypto issue.

Reviewer 2: IMO RHEL policy should have nothing to do with this. I think we are all in agreement to move the RHEL branches out of this repo so these must be upstream policies only.

Author: Still, I think I should weaken this to say "almost always". We occasionally do need to bump for CVEs and such.

Reviewer 3: FWIW, fedora reviews have [must and should](https://docs.fedoraproject.org/en-US/packaging-guidelines/ReviewGuidelines/#_things_to_check_on_review) . So, I'd say "should" is already kinda weak. But I'm cool either way. We could also account for special cases here: `except where the situation demands, for e.g. CVE fixes`.

### Tests

Please check tests added by the PR.
If a PR has no new tests, determine if this is actually appropriate - has new functionality been added which should have been tested?
If the PR does have tests, check to see if they are reasonably comprehensive.
The Podman Project does not have code coverage standards at present, but we aim to ensure that all new functionality is tested.
Tests for bugs and new functionality should, generally speaking, fail when run against Podman without the patch applied.
Reviewers are encouraged to check this when reviewing a pull request.

### Documentation

All changes to public-facing APIs (e.g. the Podman REST API) and CLI should be appropriately documented.
API changes require Swagger documentation.
CLI changes should be documented in the Manpages.

Please validate that PRs making such changes include appropriate documentation.
Many repositories in the Podman project will enforce this via CI check, but you should still review the contents of the documentation to make sure they are appropriate and complete.

## Things to Avoid

### Bikeshedding and excessively critical reviews

Please avoid bikeshedding during reviews.

> Trivial changes that do not affect the ultimate functionality of the PR - for example, small grammar or phrasing changes in documentation, unnecessary renaming of variables - should not block merge of a PR.

Reviewer 6: I agree with non-blocking PRs for internal code comments, but I disagree with anything that is customer-facing, whether it's output to be read by a customer or, more importantly, the man pages. Those errors in documentation should be blockable, IMO.

Reviewer 3: Would be good to make spellcheck or other tools a part of CI (I think we already do it in some projects). While they need not be PR blockers, I'd say it reflects poorly if we have simple typos in documentation. While we're at it, maybe we should also stick with a variant of English unless chosen already. For example, Fedora [enforces American English](https://docs.fedoraproject.org/en-US/packaging-guidelines/ReviewGuidelines/#_things_to_check_on_review): 
```
MUST: The spec file must be written in American English. See Packaging Guidelines: Summary and description
```

It is acceptable to make such comments, but they should be marked as nice to have changes.

### Asking for unrelated changes

Please do not request that a contributor make significant changes to code their PR did not touch.
If you are reviewing and find problems in pre-existing code in a file that the PR changed, you should not require that the contributor change this code as well.

> Asking if they are willing to do so is fine, but do not block merge of the PR if they are unwilling.

Reviewer 6: ```suggestion
Asking if they are willing to do so is fine, but do not block the merge of the PR if they are unwilling.
```

Being respectful includes respect for the contributor's time.
If there is a disagreement, please do not make the contributor make repeated changes until an agreement on how to proceed is reached.
Also, please attempt to reach an agreement quickly, so the PR can be merged in a timely fashion.

### Language Version Updates

Most repositories in the Podman project are written in Go, and as such target a specific version of Go in their `go.mod`.
For example:
```
$ cat go.mod | grep 'go 1.'
go 1.22.8
```

Changing this value affects what versions of Go can build the project, and as such increasing it can prevent some distributions with older Golang versions from building Podman.

For all branches except the main branch, Go version should remain static unless there is an extremely good reason to change it (for example, a CVE fix requires pulling in a new version of a library that needs a newer Golang version).
PRs into a non-main branch which change the supported Go version should be modified to not require such a change, or rejected if that is not possible.
Changing supported Go version in the main branch is allowed, but not encouraged.

> ### Tests

Reviewer 8: Documentation changes and certain changes don't require `tests` for such cases `No New test` label is used, should it reflect here ?

It is acceptable to make such comments, but they should be marked as nice to have changes.
Exceptions can be made with documentation, as ensuring correctness and clarity in documentation is very important.
Ensuring our documentation is free of typos and obvious grammatical issues is not bikeshedding and is allowed.

### Asking for unrelated changes

Please do not request that a contributor make significant changes to code their PR did not touch.
If you are reviewing and find problems in pre-existing code in a file that the PR changed, you should not require that the contributor change this code as well.
Asking if they are willing to do so is fine, but do not block the merge of the PR if they are unwilling.

Some changes are OK to ask for - for example, asking a contributor to refactor existing code very similar to something being added to prevent code duplication.
However, larger changes that would substantially increase the size of the PR should be avoided.
Reviewers should use their best judgement to balance respect for the contributor's time and the code hygiene of the project.

> If a change is too large to be reasonably asked for, consider asking the contributor to add a comment with a "TODO" or "FIXME" to the area that needs changing (or making a PR yourself with such a comment).

Reviewer 9: I’m learning a bit more towards the “let the PR contributor do all the work” side, at least in the sense that letting a drive-by contributor add a badly-structured workaround or a kludgy minimal implementation of a feature, “succeed”, and disappear, even if accompanied by a FIXME documenting this, is not a good trade for the long-term maintainability of the project.

The only tools to manage maintainer overload we have is:
- finding new contributors, and
- backpressure: not encouraging/accepting contributions that increase maintainer overload.

And there’s a trade-off between being welcoming and encouraging to new long-term contributors, vs. discouraging drive-by tech debt additions, and it’s hard to tell in advance which one we are dealing with.

Ultimately, I don’t have any specific wording suggestions, no objections to “best judgement” — and, anyway, this PR already includes a process to resolve disagreements between reviewers/maintainers.

Reviewer 9: E.g. researching a situation, documenting it, and adding a FIXME, without actually fixing it, can certainly be a valuable contribution alone.

Adding a hack that works 40% of the time, and a FIXME to turn that into a production feature, maybe not?

---

One example that I think happens somewhat regularly, incl. by regular contributors, is adding a Podman feature which doesn’t work `podman-remote`: just not doing the HTTP work when it is clearly possible, or, in situations where it’s unclear whether it is possible or where it seems difficult, not arriving at a plan and merging the local-only feature anyway.

I think the past practice for this has been “use best judgement”, and that’s fine. Just something to point at / think about.

Author: My basic intent here was to discourage asking drive-by contributors to refactor massive sizable amounts of the program (that they likely have no direct knowledge of) to improve our tech-debt situation. That kind of thing is definitely valuable, but should be done by actual maintainers or willing volunteers; we shouldn't be asking someone trying to submit a trivial patch to do it.

On the other hand, the points you make are quite valid. We should have the ability to reject contributions on grounds they will introduce excessive complexity or tech debt, and asking for a refactor of that is quite fine.

I'll try and think of a better way to phrase this. One thing I might include is that maintainers should be held to a higher standard - I'm a lot more forgiving of asking someone deeply familiar with the project to refactor than I am a drive-by contributor, and I don't think that's controversial.

Reviewer 9: > discourage asking drive-by contributors to refactor massive amounts of the program 

Agreed.

Overall I think we are on the same page, and I can’t think of any obvious wording improvement — so maybe it’s not worth spending a lot of effort trying to polish this to perfection. “There are tradeoffs, they are not 100% clear-cut, just be aware of the principal concerns.” is fine.
